tune-threads/tune-threads.lisp

```diff
@@ -65,16 +63,15 @@
 
 (defvar *default-testing-time-limit* 2.0)
 
-(defun scan-num-threads (&key (time-limit *default-testing-time-limit*) (num-qubits 12)
-                              (max-num-threads (qvm:count-logical-cores))
-                              (min-num-threads 1)
-                              (q (prepare-hadamard-test num-qubits)))
+(defun scan-num-threads (num-qubits &key (time-limit *default-testing-time-limit*)
+                                          (min-num-threads 1) (max-num-threads (qvm:count-logical-cores))
+                                          (q (prepare-hadamard-test num-qubits)))
   "Return a list of execution time per sample per thread of the qvm/program Q for different numbers of threads.
 The number of threads varies from MIN-NUM-THREADS to MAX-NUM-THREADS."
   (norm-min
-   (loop :for num-trials := (find-num-trials q num-threads time-limit 0)
-           :then (find-num-trials q num-threads time-limit num-trials)
-         :for num-threads :from min-num-threads :to max-num-threads
+   (loop :for num-threads :from min-num-threads :to max-num-threads
+         :for num-trials := (find-num-trials q num-threads time-limit)
+           :then (if (> num-trials 1) (find-num-trials q num-threads time-limit) 1)
```

Comment:
For small `num-qubits` the trial program will run thousands of times per second.  `find-num-trials` repeatedly doubles the number of trials and measures the time until the time is greater than `0.1`s, computing trials/s from the last measurement. From this value of trial/s it computes how many trials can be completed in `time-limit`s. Yes, the number of times per second decreases roughly monotonically. Eventually, `find-num-trials` will expend more time finding the number of trials than `scan-num-threads` spends running the trials. I do not approach this point efficiently, but a  zeroeth order optimization is to stop looking for the optimum number of trials when you know it is not more than one.

Reply:
I see. Thanks for that explanation. I was imagining the opposite case of a slow, complex program on many qubits, where `find-num-trials` returns 1 the first time through the loop, when `num-threads` = 1, then considered what would happen as `num-threads` increases from 1 to # logical cores. If parallelization helps your program, I'd expect `num-trials` to go up as runtime goes down, except it can't because it's locked at 1. But I guess such small changes in `num-trials` for slow programs are not that important, and the common case is `num-trials` starting high and decreasing as `num-threads` goes up.